src/ast.c

```diff
@@ -800,15 +800,22 @@ static jl_value_t *copy_ast(jl_value_t *expr, jl_svec_t *sp, int do_sp)
     }
     else if (jl_is_lambda_info(expr)) {
         jl_lambda_info_t *li = (jl_lambda_info_t*)expr;
-        /*
-        if (sp == jl_empty_svec && li->ast &&
-            jl_array_len(jl_lam_capt((jl_expr_t*)li->ast)) == 0)
+        if (sp == jl_emptysvec && li->ast &&
+            (jl_is_expr(li->ast) ? jl_array_len(jl_lam_capt((jl_expr_t*)li->ast)) == 0 : li->capt == NULL)) {
+            // share the inner function if the outer function has no sparams to insert
+            if (!li->specTypes) {
+                // if the decl values haven't been evaluated yet (or alternately already compiled by jl_trampoline), do so now
+                li->ast = jl_prepare_ast(li, li->sparams);
+                jl_gc_wb(li, li->ast);
+                if (jl_array_len(jl_lam_staticparams(li->ast)) == 0)
+                    // mark this as compilable; otherwise, will need to make an (un)specialized version of
+                    // to handle all of the static parameters before compiling
+                    li->specTypes = jl_anytuple_type; // no gc_wb needed
```

Comment:
Should expand the comment in julia.h to explain this extra implication of the `specTypes` field.